File: SW1/2/src/UserManager/user_factory.py
from models.user import Admin, Student, Teacher


class UserFactory:
    def create_user(self, username, password, email, user_id, role):
        role = role.lower()

        if role == "student":
            return Student(username, password, email, user_id)
        elif role == "teacher":
            return Teacher(username, password, email, user_id)
        elif role == "admin":
            return Admin(username, password, email, user_id)
        else:
            raise ValueError("Invalid role specified")

File: SW1/csharp/2/UserFactory.cs
namespace CourseManagementSystem;

public sealed class UserFactory
{
    public User CreateUser(string username, string password, string email, string userId, string role)
    {
        return role.Trim().ToLowerInvariant() switch
        {
            "student" => new Student(username, password, email, userId),
            "teacher" => new Teacher(username, password, email, userId),
            "admin" => new Admin(username, password, email, userId),
            _ => throw new InvalidOperationException("Invalid role specified"),
        };
    }
}
